File: examples/moss-router/lib/pricing.js
/**
 * pricing.js — static price table + cost calc (June 2026 prices, per 1M tokens)
 * All non-deprecated. Includes Groq LPU models.
 */

const DEFAULT_PRICES = {
  // OpenAI
  'gpt-4.1-nano': { in: 0.10, out: 0.40 },
  'gpt-4.1-mini': { in: 0.40, out: 1.60 },
  'gpt-4.1': { in: 2.00, out: 8.00 },
  'gpt-4o': { in: 2.50, out: 10.00 },
  'o3': { in: 2.00, out: 8.00 },
  'gpt-5.5': { in: 5.00, out: 30.00 },

  // Anthropic (current series)
  'claude-haiku-4.5': { in: 0.90, out: 4.50 },
  'claude-sonnet-4.6': { in: 3.00, out: 15.00 },
  'claude-opus-4.8': { in: 5.00, out: 25.00 },

  // Google Gemini
  'gemini-2.0-flash': { in: 0.10, out: 0.40 },
  'gemini-2.5-flash': { in: 0.30, out: 2.50 },
  'gemini-2.5-pro': { in: 1.25, out: 10.00 },

  // Groq (LPU - very cheap + fast)
  'llama-3.1-8b-instant': { in: 0.05, out: 0.08 },
  'openai/gpt-oss-20b': { in: 0.075, out: 0.30 },
  'meta-llama/llama-4-scout-17b-16e-instruct': { in: 0.11, out: 0.34 },
  'openai/gpt-oss-120b': { in: 0.15, out: 0.60 },
  'qwen/qwen3-32b': { in: 0.29, out: 0.59 },
  'llama-3.3-70b-versatile': { in: 0.59, out: 0.79 },
  'qwen/qwen3.6-27b': { in: 0.60, out: 3.00 },

  // OpenRouter pass-through examples (approx)
  'deepseek/deepseek-v4-flash': { in: 0.12, out: 0.25 },
};

export function getPrice(model) {
  if (!model) return { in: 0.5, out: 1.5 }; // safe fallback
  // normalize common variants
  const m = String(model).toLowerCase();
  if (DEFAULT_PRICES[m]) return DEFAULT_PRICES[m];
  // prefix match for groq names etc
  for (const [k, p] of Object.entries(DEFAULT_PRICES)) {
    if (m.includes(k) || k.includes(m.split('/').pop())) return p;
  }
  return { in: 0.5, out: 1.5 };
}

export function calculateCost(usage, modelOrPrice) {
  if (!usage) return 0;
  const p = typeof modelOrPrice === 'object' ? modelOrPrice : getPrice(modelOrPrice);
  const pin = usage.prompt_tokens || usage.input_tokens || 0;
  const pout = usage.completion_tokens || usage.output_tokens || 0;
  const cost = (pin / 1_000_000 * p.in) + (pout / 1_000_000 * p.out);
  return Number(cost.toFixed(6));
}

export const TIER_DEFAULTS = {
  nano: {
    description: 'Ultra low cost (classification, extraction, formatting)',
    models: [
      { provider: 'google', model: 'gemini-2.0-flash', priceIn: 0.10, priceOut: 0.40 },
      { provider: 'groq', model: 'llama-3.1-8b-instant', priceIn: 0.05, priceOut: 0.08 },
      { provider: 'openrouter', model: 'deepseek/deepseek-v4-flash', priceIn: 0.12, priceOut: 0.25 }
    ]
  },
  eco: {
    description: 'Balanced daily driver',
    models: [
      { provider: 'groq', model: 'openai/gpt-oss-20b', priceIn: 0.075, priceOut: 0.30 },
      { provider: 'google', model: 'gemini-2.5-flash', priceIn: 0.30, priceOut: 2.50 },
      { provider: 'openai', model: 'gpt-4.1-mini', priceIn: 0.40, priceOut: 1.60 }
    ]
  },
  standard: {
    description: 'Most work (good quality/cost)',
    models: [
      { provider: 'openai', model: 'gpt-4.1', priceIn: 2.00, priceOut: 8.00 },
      { provider: 'anthropic', model: 'claude-sonnet-4.6', priceIn: 3.00, priceOut: 15.00 },
      { provider: 'groq', model: 'llama-3.3-70b-versatile', priceIn: 0.59, priceOut: 0.79 }
    ]
  },
  premium: {
    description: 'Complex reasoning',
    models: [
      { provider: 'groq', model: 'openai/gpt-oss-120b', priceIn: 0.15, priceOut: 0.60 },
      { provider: 'openai', model: 'gpt-4o', priceIn: 2.50, priceOut: 10.00 },
      { provider: 'anthropic', model: 'claude-sonnet-4.6', priceIn: 3.00, priceOut: 15.00 }
    ]
  },
  flagship: {
    description: 'Max quality',
    models: [
      { provider: 'openai', model: 'gpt-5.5', priceIn: 5.00, priceOut: 30.00 },
      { provider: 'anthropic', model: 'claude-opus-4.8', priceIn: 5.00, priceOut: 25.00 },
      { provider: 'groq', model: 'qwen/qwen3.6-27b', priceIn: 0.60, priceOut: 3.00 }
    ]
  }
};

export function getTierConfig(tierName, overrideConfig) {
  const base = TIER_DEFAULTS[tierName] || TIER_DEFAULTS.eco;
  if (!overrideConfig || !overrideConfig.tiers || !overrideConfig.tiers[tierName]) {
    return base;
  }
  // allow override merge
  return { ...base, ...overrideConfig.tiers[tierName] };
}

File: lib/llm_router/pricing.py
"""Pricing for internal router.

Seeded from memory/research/*/prices-2026-06-27.json (OpenRouter/Groq 2026-06-27).
Free models have 0 cost. Minimal subset + defaults for used groups.
"""
from __future__ import annotations

from typing import Any

# Prices per 1M tokens (input, output). 0 for free/:free .
PRICING: dict[str, tuple[float, float]] = {
    # Free OR (from table 2026-06-27)
    "liquid/lfm-2.5-1.2b-instruct:free": (0.0, 0.0),
    "liquid/lfm-2.5-1.2b-thinking:free": (0.0, 0.0),
    "cohere/north-mini-code:free": (0.0, 0.0),
    "google/gemma-4-31b-it:free": (0.0, 0.0),
    "meta-llama/llama-3.2-3b-instruct:free": (0.0, 0.0),
    "openrouter/free": (0.0, 0.0),
    # Groq free-ish / compound (prices null -> treat as 0 for router)
    "groq/compound": (0.0, 0.0),
    "groq/compound-mini": (0.0, 0.0),
    # Groq nano/eco (from groq prices)
    "llama-3.1-8b-instant": (0.05, 0.08),
    "openai/gpt-oss-20b": (0.075, 0.30),
    "llama-3.3-70b-versatile": (0.59, 0.79),
    "openai/gpt-oss-120b": (0.15, 0.60),
    # Cheap paid OR (from prices)
    "inclusionai/ling-2.6-flash": (0.01, 0.03),
    # Others used in groups (approx from research)
    "openrouter/deepseek/deepseek-v4-pro": (0.30, 0.90),  # approx
    "openrouter/qwen/qwen3.6-35b-a3b": (0.14, 1.00),
    "qwen/qwen3-coder-30b-a3b-instruct": (0.07, 0.27),
    "openrouter/x-ai/grok-4.3": (1.25, 5.00),  # rough, not cheapest
    "openrouter/qwen/qwen3-next-80b-a3b-thinking": (0.10, 0.80),
    "openrouter/anthropic/claude-sonnet-4.6": (3.00, 15.00),
    "openrouter/x-ai/grok-4.20": (1.25, 5.00),
    "default": (0.50, 1.50),
}

def get_price(model: str) -> tuple[float, float]:
    """Return (input_per_1m, output_per_1m) for model. Case-insensitive prefix match."""
    if not model:
        return PRICING["default"]
    m = model.lower().strip()
    if m in PRICING:
        return PRICING[m]
    for k, p in PRICING.items():
        if k in m or m in k or k.split("/")[-1] in m:
            return p
    return PRICING["default"]

def calculate_cost(usage: Any, model: str) -> float:
    """Calculate USD from usage (prompt_tokens, completion_tokens) or dict."""
    if usage is None:
        return 0.0
    try:
        if hasattr(usage, "prompt_tokens"):
            pt = getattr(usage, "prompt_tokens", 0) or 0
            ct = getattr(usage, "completion_tokens", 0) or 0
        elif isinstance(usage, dict):
            pt = usage.get("prompt_tokens", usage.get("input_tokens", 0)) or 0
            ct = usage.get("completion_tokens", usage.get("output_tokens", 0)) or 0
        else:
            pt = int(getattr(usage, "prompt_tokens", 0) or 0)
            ct = int(getattr(usage, "completion_tokens", 0) or 0)
        pin, pout = get_price(model)
        cost = (pt / 1_000_000.0 * pin) + (ct / 1_000_000.0 * pout)
        return round(cost, 8)
    except Exception:
        return 0.0
